client/src/main/java/com/tencent/rss/client/impl/ShuffleWriteClientImpl.java

```diff
@@ -95,30 +102,50 @@ private void sendShuffleDataAsync(
       Map<ShuffleServerInfo, List<Long>> serverToBlockIds,
       Set<Long> successBlockIds,
       Set<Long> tempFailedBlockIds) {
+
+    // maintain the count of blocks that have been sent to the server
+    Map<Long, AtomicInteger> blockIdsTracker = Maps.newConcurrentMap();
+    serverToBlockIds.values().forEach(
+      blockList -> blockList.forEach(block -> blockIdsTracker.put(block, new AtomicInteger(0)))
+    );
+
     if (serverToBlocks != null) {
       serverToBlocks.entrySet().parallelStream().forEach(entry -> {
         ShuffleServerInfo ssi = entry.getKey();
         try {
+          Map<Integer, Map<Integer, List<ShuffleBlockInfo>>> shuffleIdToBlocks = entry.getValue();
+          // todo: compact unnecessary blocks that reach replicaWrite
           RssSendShuffleDataRequest request = new RssSendShuffleDataRequest(
-              appId, retryMax, retryIntervalMax, entry.getValue());
+              appId, retryMax, retryIntervalMax, shuffleIdToBlocks);
           long s = System.currentTimeMillis();
           RssSendShuffleDataResponse response = getShuffleServerClient(ssi).sendShuffleData(request);
           LOG.info("ShuffleWriteClientImpl sendShuffleData cost:" + (System.currentTimeMillis() - s));
 
           if (response.getStatusCode() == ResponseStatusCode.SUCCESS) {
-            successBlockIds.addAll(serverToBlockIds.get(ssi));
+            // mark a replica of block that has been sent
+            serverToBlockIds.get(ssi).forEach(block -> blockIdsTracker.get(block).incrementAndGet());
             LOG.info("Send: " + serverToBlockIds.get(ssi).size()
                 + " blocks to [" + ssi.getId() + "] successfully");
           } else {
-            tempFailedBlockIds.addAll(serverToBlockIds.get(ssi));
-            LOG.error("Send: " + serverToBlockIds.get(ssi).size() + " blocks to [" + ssi.getId()
+            LOG.warn("Send: " + serverToBlockIds.get(ssi).size() + " blocks to [" + ssi.getId()
                 + "] failed with statusCode[" + response.getStatusCode() + "], ");
           }
         } catch (Exception e) {
-          tempFailedBlockIds.addAll(serverToBlockIds.get(ssi));
-          LOG.error("Send: " + serverToBlockIds.get(ssi).size() + " blocks to [" + ssi.getId() + "] failed.", e);
+          LOG.warn("Send: " + serverToBlockIds.get(ssi).size() + " blocks to [" + ssi.getId() + "] failed.", e);
         }
       });
+
+      // check success and failed blocks according to the replicaWrite
+      blockIdsTracker.entrySet().forEach(blockCt -> {
```

Comment:
if we have 3 replicas, we should write 2 replicas first, and if we accounter 1 error,  we write 3th replica?

Reply:
This check is processed after writing 3 replicas are finished.

Reply:
> This check is processed after writing 3 replicas are finished.

If we wanna use this feature online, we would better have this improvement.

Reply:
I am planning to optimize this in another PR, because it needs to refactor the original RPC sending logic.

Reply:
OK.